<div style="text-align:left;font-size:2em"><span style="font-weight:bolder;font-size:1.25em">SP2273 | Learning Portfolio</span><br><br><span style="font-weight:bold;color:darkred">Loops (Good) Exercises</span></div>

# Exercise 1
With your knowledge of growing lists, use a `for` loop with `range()` and `continue` to generate a list of the squares of the odd integers from 0 to 9.

In [12]:
odd_integer_squares = []                          # Creates the empty list

for integers in range(10):                        # Iterate through integers from 1 to 10
    if integers % 2 == 0:
        continue                                  # Exclude even numbers
    odd_integer_squares.append(integers**2)       # Adds the odd integer squares to the list

print(odd_integer_squares)

[1, 9, 25, 49, 81]


# Exercise 2
Redo the previous exercise using list comprehension.

In [1]:
import numpy as np
odd_integer_squares_2 = [number**2 for number in range(10) if number % 2 == 1]
print(odd_integer_squares_2)

[1, 9, 25, 49, 81]


In [13]:
#import numpy as np
#odd_integer_list = [number for number in range(10) if number % 2 == 1]    # Creates a list of odd integers from 1 to 10
#odd_integer_array = np.array(odd_integer_list)                            # Turns this list into an array so that we can square it later
#print(odd_integer_array)

[1 3 5 7 9]


In [14]:
#odd_integer_squares_array = (odd_integer_array)**2
#print(odd_integer_squares_array)

[ 1  9 25 49 81]


# Exercise 3

In [2]:
%%timeit
odd_integer_squares = []
for integers in range(10):
    if integers % 2 == 0:
        continue
    odd_integer_squares.append(integers**2)

2.05 µs ± 66.1 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


In [3]:
%%timeit
odd_integer_squares_2 = [number**2 for number in range(10) if number % 2 == 1]

2.11 µs ± 155 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


The first method takes `2.05 µs ± 66.1 ns` per loop whereas the second takes `2.11 µs ± 155 ns` per loop. Therefore we can conclude that both methods are just as fast.

# Exercise 4
The initial quantity of a sample of a radioactive substance is 100 units, and it decays by 5% each year. Use a `while` loop to determine how long the sample will take to reduce to half its original amount.

In [4]:
years = 0
amount = 100
while amount > 50:
    amount = amount*0.95                                              # For every loop,
    years += 1                                                        # add one year
print(f'The half-life of this radioactive substance is around {years} years.')

The half-life of this radioactive substance is around 14 years.


# Exercise 5
The following is data about atmospheric $CO_2$
 levels for several years in a (year, CO2_level) format. The units are ppm.

In [6]:
co2_data = [
    (2000, 369.55), (2001, 371.14), (2002, 373.28), 
    (2003, 375.80), (2004, 377.52), (2005, 379.80), 
    (2006, 381.90), (2007, 383.79), (2008, 385.60), 
    (2009, 387.43), (2010, 389.90), (2011, 391.65), 
    (2012, 393.85), (2013, 396.52), (2014, 398.65),
    (2015, 400.83), (2016, 404.24), (2017, 406.55), 
    (2018, 408.52), (2019, 411.44), (2020, 414.24)
]

Identify those years that showed an increase of $CO_2$ of 3 ppm or more compared to the previous year.
Please print out these years along with the corresponding change in concentration.

In [8]:
current_year = 0
previous_year = -1

for year, ppm in co2_data:
    increase = co2_data[current_year][1] - co2_data[previous_year][1]                # This takes the difference between CO2 concentrations between the
    current_year += 1                                                                # last [-1] and first [1] element. I am not sure how this can be
    previous_year += 1                                                               # resolved without having an error message
    if increase < 3:
        continue
    print(f"There was a {increase:.3} ppm increase of CO2 in {year} compared to {year-1}")   # Leaves the increase to 3 s.f.

There was a 3.41 ppm increase of CO2 in 2016 compared to 2015


In [20]:
for year, co2_level in co2_data[1:]:                                                 # Skip the first year since there is no previous year
  previous_year, previous_co2_level = co2_data[year - 2000 - 1]                      # previous_year and pervious_co2_level unpack elements from one previous element
  change_in_co2 = co2_level - previous_co2_level

  if change_in_co2 >= 3:                                                             # Check if the change in CO2 is at least 3 ppm
    print(f"There was a {change_in_co2:.3} ppm increase of CO2 in {year} compared to {year-1}")   # Leaves the increase to 3 s.f.

There was a 3.41 ppm increase of CO2 in 2016 compared to 2015
